core/src/main/java/com/linecorp/armeria/common/AbstractRequestContext.java

```diff
@@ -98,33 +111,37 @@ public final Runnable makeContextAware(Runnable runnable) {
     public final <T, U> BiConsumer<T, U> makeContextAware(BiConsumer<T, U> action) {
         return (t, u) -> {
             try (SafeCloseable ignored = propagateContextIfNotPresent()) {
+                cancelIfTimedOut();
                 action.accept(t, u);
             }
         };
     }
 
     @Override
     public final <T> FutureListener<T> makeContextAware(FutureListener<T> listener) {
-        return future -> invokeOperationComplete(listener, future);
+        return future -> invokeOperationComplete(listener, future, () -> new DefaultPromise<>(eventLoop()));
     }
 
     @Override
     public final ChannelFutureListener makeContextAware(ChannelFutureListener listener) {
-        return future -> invokeOperationComplete(listener, future);
+        return future -> invokeOperationComplete(
+                listener, future, () -> new DefaultChannelPromise(future.channel()));
     }
 
     @Override
     public final <T extends Future<?>> GenericFutureListener<T>
     makeContextAware(GenericFutureListener<T> listener) {
-        return future -> invokeOperationComplete(listener, future);
+        return future -> invokeOperationComplete(listener, future, null);
     }
 
     @Override
     public final <T> CompletionStage<T> makeContextAware(CompletionStage<T> stage) {
         final CompletableFuture<T> future = new RequestContextAwareCompletableFuture<>(this);
         stage.whenComplete((result, cause) -> {
             try (SafeCloseable ignored = propagateContextIfNotPresent()) {
-                if (cause != null) {
+                if (timedOut) {
+                    future.cancel(true);
+                } else if (cause != null) {
```

Comment:
This seems somewhat awkward. Isn't the `stage` done already. I see no point of canceling `future`. It will be completed anyway.